### DISCUSSION TITLE
How to handle dirty form data?

### DISCUSSION THREAD

**Discussion by aashrafh** [remix-run/remix#4909] [Q&A]
Hello everyone :wave: 

I'm working on a Remix based project and encountered the following problem; I have a part of a form that you can add multiple languages to your profile. You can add a new language or remove an old one but these two actions are considered "a dirty state", nothing is saved to the database until you press the "Save" button.

One of the approaches to handle this, of course, is to use React's state to store the dirty language list. But, I'm wondering, is there a way to handle it without client-side JavaScript/React's state? I tried to do this by having an `isDirty` flag in the action data and created [this demo](https://codesandbox.io/p/sandbox/old-forest-x8bhus?selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A17%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A17%7D%5D&file=%2Fapp%2Froutes%2Findex.tsx). Do you have a better approach?

**Comment by sergiodxa** [comment]
If you want to show something to the user on the UI, you need JS.

If you want to know on the action, you need to wait until the user submits the form, get the data and check if it's different than the saved data.
